### DISCUSSION TITLE
Recurrence overhaul

### DISCUSSION THREAD

**Discussion by skbolton** [GothenburgBitFactory/taskwarrior#2491] [Ideas]
@tbabej go ahead and close this if you don't feel like this is the forum for discussion on the new changes being considered and drafted [here](https://taskwarrior.org/docs/design/recurrence.html#add-periodic-template)

I had one idea around the `periodic` and `chained` recurrence type that I wanted to share. In Todoist they have this syntax that you can add a`!` to the end of a recurrence to say that it only recurs when the task has been completed. I think this could work nicely in taskwarrior syntax because the old system of recurrence would look the same.

```
task add proj:Finances recur:yearly due:2022-04-15 Do Taxes
```

For the the chained type you could do this syntax.

```
task add proj:Home recur:weekly! due:sunday Mow the lawn
```

I like how lightweight this is as opposed to adding a new `rtype` flag into the command line. Also it shouldn't be too bad to parse when grabbing the recur value.

**Comment by tbabej** [comment]
:+1: This is the right place to discuss ideas like this. I'll chime in later.

**Reply by skbolton** [reply]
Already forgetting that `!` is a special character in shell so not as handy in practice. But I wonder if some sort of modification to the recur date to encompass chained could still be nice.

Maybe `recur:weekly^`

**Reply by tbabej** [reply]
I think you raise an important point. The current proposal is indeed not "quick and elegant" on the cli, which goes a bit against the spirit of TW (which for example, goes out of its way to make itself nible and quick, i.e. with attribute abbreviations on the CLI).

Some part of the pain could be alleviated if one of the recurrence types were the default (hence requiring you to explicitly specify the recurrence type only if you wanted to use the less common type. Or perhaps the default could be configurable.

Introducing some syntactic sugar as you suggest should be possible, but not exactly straightforward. We'd have to be careful with suffixing letters that could have other meaning (such as binary operators, etc.). In any case, I do think this should be _just_ syntactic sugar, i.e. the expression you propose would still set `recur:weekly rtype:chained` attributes in the data file, but be represented only by `recur:weekly^` on the CLI.
